storage/gcs/src/test/java/io/aiven/kafka/tieredstorage/storage/gcs/CredentialsBuilderTest.java

```diff
@@ -0,0 +1,105 @@
+        assertThatThrownBy(() -> CredentialsBuilder.build(true, "{}", null))
+            .isInstanceOf(IllegalArgumentException.class)
+            .hasMessage("defaultCredentials == true and credentialsJson != null cannot be simultaneously.");
+    }
+
+    @Test
+    void defaultAndPathProvided() {
+        assertThatThrownBy(() -> CredentialsBuilder.build(true, null, "file.json"))
+            .isInstanceOf(IllegalArgumentException.class)
+            .hasMessage("defaultCredentials == true and credentialsPath != null cannot be simultaneously.");
+    }
+
+    @Test
+    void bothCredentialsProvided() {
+        assertThatThrownBy(() -> CredentialsBuilder.build(false, "{}", "file.json"))
+            .isInstanceOf(IllegalArgumentException.class)
+            .hasMessage("Both credentialsPath and credentialsJson cannot be non-null.");
+    }
+
+    @Test
+    void defaultCredentials() throws IOException {
+        final GoogleCredentials mockCredentials = GoogleCredentials.newBuilder().build();
+        try (final MockedStatic<GoogleCredentials> googleCredentialsMockedStatic =
+                 Mockito.mockStatic(GoogleCredentials.class)) {
+            googleCredentialsMockedStatic.when(GoogleCredentials::getApplicationDefault).thenReturn(mockCredentials);
+            final Credentials r = CredentialsBuilder.build(true, null, null);
+            assertThat(r).isSameAs(mockCredentials);
+        }
+    }
+
+    @Test
+    void noCredentials() throws IOException {
+        final Credentials r = CredentialsBuilder.build(false, null, null);
+        assertThat(r).isSameAs(NoCredentials.getInstance());
+    }
+
+    @Test
+    void fromJson() throws IOException {
+        final String credentialsJson = Resources.toString(
+            Thread.currentThread().getContextClassLoader().getResource("test_gcs_credentials.json"),
+            StandardCharsets.UTF_8);
+        final Credentials credentials = CredentialsBuilder.build(false, credentialsJson, null);
+        assertCredentials(credentials);
+    }
+
+    @Test
+    void fromPath() throws IOException {
+        final String credentialsPath = Thread.currentThread()
+            .getContextClassLoader()
+            .getResource("test_gcs_credentials.json")
+            .getPath();
+        final Credentials credentials = CredentialsBuilder.build(false, null, credentialsPath);
+        assertCredentials(credentials);
+    }
+
+    private static void assertCredentials(final Credentials credentials) {
+        assertThat(credentials).isInstanceOf(UserCredentials.class);
+        final UserCredentials userCredentials = (UserCredentials) credentials;
+        assertThat(userCredentials)
+            .extracting("clientId")
+            .isEqualTo("test-client-id");
+        assertThat(userCredentials)
+            .extracting("clientSecret")
+            .isEqualTo("test-client-secret");
+        assertThat(userCredentials)
+            .extracting("refreshToken")
+            .isEqualTo("x");
```

Comment:
Since you are anyways doing the cast and there are getters, wouldn't it be better to just 
```
        assertThat(userCredentials.getClientId()).isEqualTo("test-client-id");
        assertThat(userCredentials.getClientSecret()).isEqualTo("test-client-secret");
        assertThat(userCredentials.getRefreshToken()).isEqualTo("x");
```
This also avoid unnecessary reflection calls and if the API will be changed will fail in compile time.

Reply:
Makes sense. I used to have some private stuff here in the beginning, but not now. Fixed.